File: lib/max_range_sum.rb
def max_range_sum(array)

    # Track two max's
    max = nil # => Overall maximum
    current_max = 0 # => current contiguous maximum

    array.each do |el|
        current_max += el # => add to contiguous max
        current_max = 0 if current_max < 0 # => negative contiguous max to 0 if it is negative
        max = current_max if max.nil? || max < current_max # => contiguous max becomes max if it is larger the the overall max
    end

    max

    # TIME COMPLEXITY => O(N), Linear , N being total number of elements in the array
    # SPACE COMPLEXITY => O(1), Constant, as the space used is not dependent on the input
end

File: 2017_12_12/max_range_sum.js
function maxRangeSum(arr) {

    // Tracking two max's,
    let max = null; // => Overall max
    let current_max = 0; // => Contiguous Max

    arr.forEach(function(el){
        current_max += el;
        if (current_max < 0){
            current_max = 0; // => Reset contiguous max if it is negative
        }
        if ( max === null || max < current_max ){
            max = current_max; // => Replace overall max if it is smaller than contiguous max
        }
    })
    return max;
}